Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -507,6 +529,162 @@ private byte[] GetKey(ShortHash hash)
             return hash.ToByteArray();
         }
 
+        #region Memoization
+        // TODO(jubayard): garbage collection / removal in general
+
+        /// <inheritdoc />
+        public override GetContentHashListResult GetContentHashList(OperationContext context, StrongFingerprint strongFingerprint)
+        {
+            var key = GetCaChaaSKey(strongFingerprint);
+            ContentHashListWithDeterminism? result = null;
+            var status = _keyValueStore.Use(
+                store =>
+                {
+                    if (store.TryGetValue(key, out var data, nameof(Columns.CaChaaS)))
+                    {
+                        result = DeserializeContentHashList(data);
+                        // TODO(jubayard): since we are inside the ContentLocationDatabase, we can validate that all
+                        // hashes exist. Moreover, we can prune content.
+                    }
+                });
+
+            if (!status.Succeeded)
+            {
+                return new GetContentHashListResult(status.Failure.CreateException());
+            }
+            
+            return new GetContentHashListResult(result ?? default);

```

Comment:
Maybe instead of using nullable value you can just use:

`ContentHashListWithDeterminism result = new ContentHashListDeterminism();`

Reply:
Or maybe we should be more explicit in case of a miss and use something like:

if (result == null) {
    // miss
    return new GetContentHashListResult(null, CacheDeterminism.None);
}


---
In reply to: [302175997](https://github.com/microsoft/BuildXL/pull/571#discussion_r302175997) [](ancestors = 302175997)